tests/posix/env/src/env.c

```diff
@@ -0,0 +1,189 @@
+ */
+
+#if !defined(_POSIX_C_SOURCE) || _POSIX_C_SOURCE < 200112L
+#undef _POSIX_C_SOURCE
+#define _POSIX_C_SOURCE 200112L
+#endif
+
+#ifndef _BSD_VISIBLE
+#define _BSD_VISIBLE
+#endif
+
+#include <stdio.h>
+#include <stdlib.h>
+
+#include <zephyr/ztest.h>
+
+#define _m_home "/home/zephyr"
+#define _m_uid  STRINGIFY(1000)
+#define _m_pwd  "/tmp"
+
+#define _m_alt_home "/this/path/is/much/longer/than" _m_home
+
+#define DEFINE_ENVIRON(_handle, _key, _val) char _handle[] = STRINGIFY(_key) "=" _val
+#define RESET_ENVIRON(_handle, _key, _val)                                                         \
+	snprintf(_handle, ARRAY_SIZE(_handle), "%s=%s", _key, _val)
+
+#if defined(CONFIG_NEWLIB_LIBC) || defined(CONFIG_PICOLIBC)
+/* newlib headers seem to be missing this */
+int getenv_r(const char *name, char *val, size_t len);
+#endif
+
+extern char **environ;
+static char **old_environ;
+
+static DEFINE_ENVIRON(home, "HOME", _m_home);
+static DEFINE_ENVIRON(uid, "UID", _m_uid);
+static DEFINE_ENVIRON(pwd, "PWD", _m_pwd);
+
+static char *environ_for_test[] = {home, uid, pwd, NULL};
+
+ZTEST(env, test_getenv)
+{
+	zassert_equal(getenv(NULL), NULL);
+	zassert_equal(getenv(""), NULL);
+	zassert_equal(getenv("invalid=key"), NULL);
+	zassert_equal(getenv("HOME=" _m_home), NULL);
+	zassert_equal(getenv("PWDR"), NULL);
+
+	zassert_mem_equal(getenv("HOME"), _m_home, strlen(_m_home) + 1);
+	zassert_mem_equal(getenv("UID"), _m_uid, strlen(_m_uid) + 1);
+	zassert_mem_equal(getenv("PWD"), _m_pwd, strlen(_m_pwd) + 1);
+}
+
+ZTEST(env, test_getenv_r)
+{
+	static char buf[16];
+	static const int exp_errno[] = {
+		EINVAL, EINVAL, EINVAL, EINVAL, ENOENT, ENOENT, ENOENT, EINVAL, EINVAL, EINVAL,
+	};
+	static const struct args_s {
+		const char *name;
+		char *buf;
+		size_t size;
+	} args[] = {
+		/* invalid input */
+		{NULL, NULL, 0},
+		{NULL, NULL, 42},
+		{NULL, buf, 0},
+		{NULL, buf, sizeof(buf)},
+		{"hello", NULL, 0},
+		{"hello", NULL, 42},
+		{"hello", buf, 0},
+
+		/* invalid names */
+		{"", buf, sizeof(buf)},
+		{"invalid=key", buf, sizeof(buf)},
+		{"HOME=", buf, sizeof(buf)},
+	};
+
+	BUILD_ASSERT(ARRAY_SIZE(exp_errno) == ARRAY_SIZE(args));
+
+	ARRAY_FOR_EACH(args, i)
+	{
+		errno = 0;
+		zassert_equal(getenv_r(args[i].name, args[i].buf, args[i].size), -1,
+			      "getenv_r(\"%s\", %p, %zu): expected to fail", args[i].name,
+			      args[i].buf, args[i].size);
+		zassert_equal(errno, exp_errno[i],
+			      "getenv_r(\"%s\", %p, %zu): act_errno: %d exp_errno: %d",
+			      args[i].name, args[i].buf, args[i].size, errno, exp_errno[i]);
+	}
+
+	zassert_mem_equal(getenv("HOME"), _m_home, strlen(_m_home) + 1);
+	zassert_mem_equal(getenv("UID"), _m_uid, strlen(_m_uid) + 1);
+	zassert_mem_equal(getenv("PWD"), _m_pwd, strlen(_m_pwd) + 1);
+}
+
+ZTEST(env, test_setenv)
+{
+	zassert_equal(setenv(NULL, NULL, 0), -1);
+	zassert_equal(errno, EINVAL);
+
+	/*
+	 * bug in picolibc / newlib
+	 * https://github.com/picolibc/picolibc/issues/648
```

Comment:
Closed now (thanks @keith-packard 😁) but it might take a bit of time before the module is updated in `west.yml`